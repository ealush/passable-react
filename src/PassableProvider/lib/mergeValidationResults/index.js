import { generateFieldValidationSummary, getDefaultState } from '../index';
import deepassign from '@fiverr/futile/lib/deepassign';

export default function mergeValidationResults(state, passableObject) {
    let nextState = deepassign({}, getDefaultState(), state);
    if (!state || !passableObject) { return nextState; }

    const validationResult = deepassign({
        validationErrors: {},
        validationWarnings: {}
    }, passableObject);

    const tested = validationResult.testsPerformed || [];
    const fields = nextState.fields;

    nextState = Object.keys(tested).reduce((accumulator, current) => {
        const curr = tested[current];
        const {
            hasError,
            hasWarning
        } = generateFieldValidationSummary(curr);

        hasError ? accumulator.errors[current] = curr.failCount
            : delete accumulator.errors[current];

        hasWarning ? accumulator.warnings[current] = curr.warnCount
            : delete accumulator.warnings[current];

        accumulator.fields[current] = Object.assign({}, fields[current], {
            hasError,
            hasWarning,
            errors: validationResult.validationErrors[current] || [],
            warnings: validationResult.validationWarnings[current] || []
        });
        return accumulator;
    }, { fields, errors: nextState.errors, warnings: nextState.warnings });
    nextState.warningCount = Object.keys(nextState.warnings).length;
    nextState.errorCount = Object.keys(nextState.errors).length;
    return nextState;
}