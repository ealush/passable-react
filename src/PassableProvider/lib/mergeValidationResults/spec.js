import mergeValidationResults from './index';

describe('Test mergeValidationResults function', () => {

    const state = {
        fields: {
            field_1: {
                touched: true
            },
            field_2: {
                value: 1
            }
        }
    };

    it('should merge validation results into existing state', () => {
        expect(mergeValidationResults(state, {
            testsPerformed: {
                field_1: {
                    failCount: 1,
                    warnCount: 0,
                    testCount: 1
                }
            },
            validationErrors: {
                field_1: ['error_string']
            },
            validationWarnings: {}
        })).toMatchSnapshot();
    });

    it('should default to empty passable object when of invalid type', () => {
        expect(mergeValidationResults(state, 1)).toMatchSnapshot();;
        expect(mergeValidationResults(state, null)).toMatchSnapshot();;
        expect(mergeValidationResults(state, [])).toMatchSnapshot();;
        expect(mergeValidationResults(state, true)).toMatchSnapshot();;
        expect(mergeValidationResults(state, 'false')).toMatchSnapshot();;
        expect(mergeValidationResults(state, new Function())).toMatchSnapshot();;
    });

    it('should return default `state` if missing passable object', () => {
        const expectedResult = Object.assign(state, {
            errors: {},
            warnings: {}
        });

        expect(mergeValidationResults(state)).toMatchSnapshot();;
    });

    it('should return default `state` if no arguments passed', () => {
        expect(mergeValidationResults()).toMatchSnapshot();;
    });
});