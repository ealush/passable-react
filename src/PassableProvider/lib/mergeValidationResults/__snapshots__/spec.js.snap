// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test mergeValidationResults function should default to empty passable object when of invalid type 1`] = `
Object {
  "errorCount": 0,
  "errors": Object {},
  "fields": Object {
    "field_1": Object {
      "touched": true,
    },
    "field_2": Object {
      "value": 1,
    },
  },
  "warningCount": 0,
  "warnings": Object {},
}
`;

exports[`Test mergeValidationResults function should default to empty passable object when of invalid type 2`] = `
Object {
  "errorCount": 0,
  "errors": Object {},
  "fields": Object {
    "field_1": Object {
      "touched": true,
    },
    "field_2": Object {
      "value": 1,
    },
  },
  "warningCount": 0,
  "warnings": Object {},
}
`;

exports[`Test mergeValidationResults function should default to empty passable object when of invalid type 3`] = `
Object {
  "errorCount": 0,
  "errors": Object {},
  "fields": Object {
    "field_1": Object {
      "touched": true,
    },
    "field_2": Object {
      "value": 1,
    },
  },
  "warningCount": 0,
  "warnings": Object {},
}
`;

exports[`Test mergeValidationResults function should default to empty passable object when of invalid type 4`] = `
Object {
  "errorCount": 0,
  "errors": Object {},
  "fields": Object {
    "field_1": Object {
      "touched": true,
    },
    "field_2": Object {
      "value": 1,
    },
  },
  "warningCount": 0,
  "warnings": Object {},
}
`;

exports[`Test mergeValidationResults function should default to empty passable object when of invalid type 5`] = `
Object {
  "errorCount": 0,
  "errors": Object {},
  "fields": Object {
    "field_1": Object {
      "touched": true,
    },
    "field_2": Object {
      "value": 1,
    },
  },
  "warningCount": 0,
  "warnings": Object {},
}
`;

exports[`Test mergeValidationResults function should default to empty passable object when of invalid type 6`] = `
Object {
  "errorCount": 0,
  "errors": Object {},
  "fields": Object {
    "field_1": Object {
      "touched": true,
    },
    "field_2": Object {
      "value": 1,
    },
  },
  "warningCount": 0,
  "warnings": Object {},
}
`;

exports[`Test mergeValidationResults function should merge validation results into existing state 1`] = `
Object {
  "errorCount": 1,
  "errors": Object {
    "field_1": 1,
  },
  "fields": Object {
    "field_1": Object {
      "errors": Array [
        "error_string",
      ],
      "hasError": true,
      "hasWarning": false,
      "touched": true,
      "warnings": Array [],
    },
    "field_2": Object {
      "value": 1,
    },
  },
  "warningCount": 0,
  "warnings": Object {},
}
`;

exports[`Test mergeValidationResults function should return default \`state\` if missing passable object 1`] = `
Object {
  "errorCount": 0,
  "errors": Object {},
  "fields": Object {
    "field_1": Object {
      "touched": true,
    },
    "field_2": Object {
      "value": 1,
    },
  },
  "warningCount": 0,
  "warnings": Object {},
}
`;

exports[`Test mergeValidationResults function should return default \`state\` if no arguments passed 1`] = `
Object {
  "errorCount": 0,
  "errors": Object {},
  "fields": Object {},
  "warningCount": 0,
  "warnings": Object {},
}
`;
